src/workerd/api/messagechannel.h

```diff
@@ -0,0 +1,193 @@
+#pragma once
+
+#include <workerd/api/basics.h>
+#include <workerd/io/io-context.h>
+#include <workerd/jsg/jsg.h>
+#include <workerd/jsg/ser.h>
+
+namespace workerd::api {
+
+// An implementation of the Web platform standard MessagePort.
+// MessagePorts always come in pairs. When a message is posted to
+// one it is delivered to the other, and vice versa. When one port
+// is closed both ports are closed.
+class MessagePort final: public EventTarget {
+ public:
+  // There are several variations of MessageEvent implemented in
+  // the runtime. This one is specific to MessagePort.
+  class MessageEvent final: public Event {
+   public:
+    static jsg::Ref<MessageEvent> New(jsg::Lock& js,
+        jsg::JsRef<jsg::JsValue> data,
+        kj::String lastEventId,
+        jsg::Ref<MessagePort> source,
+        kj::Array<jsg::Ref<MessagePort>> ports);
+    static jsg::Ref<MessageEvent> NewError(jsg::Lock& js,
+        jsg::JsRef<jsg::JsValue> data,
+        kj::String lastEventId,
+        jsg::Ref<MessagePort> source,
+        kj::Array<jsg::Ref<MessagePort>> ports);
+
+    MessageEvent(kj::String name,
+        jsg::JsRef<jsg::JsValue> data,
+        kj::String lastEventId,
+        jsg::Ref<MessagePort> source,
+        kj::Array<jsg::Ref<MessagePort>> ports)
+        : Event(kj::mv(name)),
+          data(kj::mv(data)),
+          lastEventId(kj::mv(lastEventId)),
+          source(kj::mv(source)),
+          ports(kj::mv(ports)) {}
+
+    static jsg::Ref<MessageEvent> constructor() = delete;
+
+    jsg::JsRef<jsg::JsValue> getData(jsg::Lock& js) {
+      return data.addRef(js);
+    }
+    kj::StringPtr getLastEventId() {
+      return lastEventId;
+    };
+    jsg::Ref<MessagePort> getSource(jsg::Lock& js) {
+      return source.addRef();
+    };
+    kj::ArrayPtr<jsg::Ref<MessagePort>> getPorts(jsg::Lock& js) {
+      return ports;
+    }
+
+    JSG_RESOURCE_TYPE(MessageEvent) {
+      JSG_INHERIT(Event);
+      JSG_READONLY_PROTOTYPE_PROPERTY(data, getData);
+      JSG_READONLY_PROTOTYPE_PROPERTY(lastEventId, getLastEventId);
+      JSG_READONLY_PROTOTYPE_PROPERTY(source, getSource);
+      JSG_READONLY_PROTOTYPE_PROPERTY(ports, getPorts);
+      // The standard also defines the origin property, but we don't
+      // implement origin in this case so we leave it out.
+    }
+
+   private:
+    jsg::JsRef<jsg::JsValue> data;
+    kj::String lastEventId;
```

Comment:
Nit: why not just have this be a JsString?
Just for the good measure of trying to remove kj types from our JSG resources

Reply:
Yeah, I'm coming along more and more to the idea of this more and more...